configure.go

```diff
@@ -0,0 +1,146 @@
+package configure
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+)
+
+const (
+	structTagKey             = "config"
+	requiredTagKey           = "required"
+	missingValuesErrTemplate = "Missing required fields: %s"
+)
+
+var (
+	ErrStringsOnly            = errors.New("Only string values are allowed in a config struct.")
+	ErrNotReference           = errors.New("The config struct must be a pointer.")
+	ErrStructOnly             = errors.New("Config object must be a struct.")
+	ErrNoTagValue             = errors.New("Config object attributes must have a 'config' tag value.")
+	ErrTooManyTagValues       = errors.New("Config object attributes can only have a key and optional required attribute.")
+	ErrFlagParsed             = errors.New("The flag library cannot be used in conjunction with configure")
+	ErrInvalidJSON            = errors.New("Invalid JSON found in arguments.")
+	ErrStructTagInvalidOption = errors.New("Only 'required' is a config option.")
+)
+
+func parseTagKey(tag string) (key string, required bool, err error) {
+	if tag == "" {
+		return "", false, ErrNoTagValue
+	}
+
+	s := strings.Split(tag, ",")
+	switch len(s) {
+	case 2:
+		if s[1] != requiredTagKey {
+			return "", false, ErrStructTagInvalidOption
+		}
+		return s[0], true, nil
+	case 1:
+		return s[0], false, nil
+	default:
+		return "", false, ErrTooManyTagValues
+	}
+}
+
+// Configure takes a reference to an interface that has 'config' tags on all atttributes of
+// the struct. Configure first tries to find values for these attributes through command line
+// flags, then will attempt to parse the first argument as a JSON blob.
+// An attribute can be required by appending ',required' to the config key.
+func Configure(configStruct interface{}) error {
+	if flag.Parsed() {
+		return ErrFlagParsed
+	}
+
+	reflectConfig := reflect.ValueOf(configStruct)
```

Comment:
Is there anyway to get around using reflect with a type switch? Also, if we're erroring on it not being a pointer, could we be more specific about this in the error message or in a comment?